src/proxy/NativeMessagingProxy.cpp

```diff
@@ -0,0 +1,107 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "NativeMessagingProxy.h"
+#include "browser/BrowserShared.h"
+
+#include <QCoreApplication>
+#include <QtConcurrent/QtConcurrent>
+
+#include <iostream>
+
+#ifdef Q_OS_WIN
+#include <fcntl.h>
+#include <windows.h>
+#endif
+
+NativeMessagingProxy::NativeMessagingProxy()
+    : QObject()
+{
+    connect(this,
+            &NativeMessagingProxy::stdinMessage,
+            this,
+            &NativeMessagingProxy::processStandardInput,
+            Qt::QueuedConnection);
+
+    setupStandardInput();
+    setupLocalSocket();
+}
+
+void NativeMessagingProxy::setupStandardInput()
+{
+#ifdef Q_OS_WIN
+    setmode(fileno(stdin), _O_BINARY);
+    setmode(fileno(stdout), _O_BINARY);
+#endif
+
+    QtConcurrent::run([this] {
+        while (std::cin.good()) {
+            if (std::cin.peek() != EOF) {
+                unsigned int length = 0;
+                for (int i = 0; i < 4; ++i) {
+                    length |= getchar() << (i * 8);
+                }
+
+                QString msg;
+                for (unsigned int i = 0; i < length; ++i) {
+                    msg += getchar();
+                }
+
+                if (msg.length() > 0) {
+                    emit stdinMessage(msg);
+                }
+            }
+            QThread::msleep(100);
+        }
+        QCoreApplication::quit();
+    });
+}
+
+void NativeMessagingProxy::processStandardInput(const QString& msg)
+{
+    if (m_localSocket && m_localSocket->state() == QLocalSocket::ConnectedState) {
+        m_localSocket->write(msg.toUtf8(), msg.length());
+        m_localSocket->flush();
+    }
+}
+
+void NativeMessagingProxy::setupLocalSocket()
+{
+    m_localSocket = new QLocalSocket();
+    m_localSocket->connectToServer(BrowserShared::localServerPath());
+    m_localSocket->setReadBufferSize(BrowserShared::NATIVEMSG_MAX_LENGTH);
+
+    connect(m_localSocket, SIGNAL(readyRead()), this, SLOT(newLocalMessage()));
+    connect(m_localSocket, SIGNAL(disconnected()), this, SLOT(deleteSocket()));
+}
+
+void NativeMessagingProxy::newLocalMessage()
+{
+    auto msg = m_localSocket->readAll();
+    if (!msg.isEmpty()) {
+        uint len = msg.size();
+        std::cout << char(((len >> 0) & 0xFF)) << char(((len >> 8) & 0xFF)) << char(((len >> 16) & 0xFF))
+                  << char(((len >> 24) & 0xFF));
+        std::cout << msg.toStdString() << std::flush;
```

Comment:
Too many magic numbers. No idea what's going on here. And why are we using `iostream` instead of `QTextStream`?

Reply:
The message length must be passed at the beginning as a 32-bit number (native byte order). If we can replace this with a QTextStream, that would be nice.

Reply:
This is the only way I can get it to work, this is from the original implementation prior to refactor.

Reply:
QTextStream should work just like std::iostream.

Reply:
This works great on all platforms and I don't really want to mess with it anymore.

Reply:
I just don't like mixing libraries too much and we're using QTextStream everywhere else. It should really be a drop-in replacement. And please give some names to the magic numbers or add a short comment above.

Reply:
They aren't magic, just bit shifting 0, 1, 2, 3 bytes 😛

Reply:
I made it sexy for yah

Reply:
I can see that you're shifting bytes, but it's not immediately obvious why and what the expected outcome is.